challengeutils/writeups.py

```diff
@@ -0,0 +1,162 @@
+    submissions = list(utils.evaluation_queue_query(
+        syn,
+        "select objectId, team from evaluation_{} "
+        "where status == 'SCORED'".format(submission_queueid)))
+    writeupsdf = pd.DataFrame(writeups)
+    submissionsdf = pd.DataFrame(submissions)
+    submissions_with_writeupsdf = \
+        submissionsdf.merge(writeupsdf, on="team", how="left")
+
+    submissions_with_writeupsdf.apply(
+        lambda row: append_writeup_to_main_submission(row, syn), axis=1)
+
+
+def validate_project(syn, submission, challenge, public=False, admin=None):
+    """
+    Validate a Project submission; errors found are returned in a dict.
+
+    Args:
+        submission - submission ID
+        challenge - Synapse ID of Challenge wiki
+        public - Project should be public (default: False)
+        admin - (optional) Project should be shared with this username/ID
+    """
+    writeup = syn.getSubmission(submission)
+    errors = []
+
+    type_error = _validate_ent_type(writeup)
+    if type_error:
+        errors.append(type_error)
+
+    contents_error = _validate_project_contents(writeup, challenge)
+    if contents_error:
+        errors.append(contents_error)
+
+    try:
+        if public:
+            errors.extend(_validate_public_permissions(syn, writeup))
+
+        if admin is not None:
+            admin_error = _validate_admin_permissions(syn, writeup, admin)
+            if admin_error:
+                errors.append(admin_error)
+
+    except SynapseHTTPError as e:
+        if e.response.status_code == 403:
+            errors.append(
+                "Submission is private; please update its sharing settings.")
+
+    status = "INVALID" if errors else "VALIDATED"
+    return {'errors_found': errors, "writeup_status": status}
+
+
+def archive_project(syn, submission, admin):
+    """
+    Make a copy (archive) of the Project submission.
+
+    Args:
+        submission - submission ID
+        admin - user who will own the archived project
+    """
+    writeup = syn.getSubmission(submission)
+    name = writeup.name.replace("&", "+").replace("'", "")
+    curr_time = int(round(time.time() * 1000))
+    new_project = Project(f"Archived {name} {curr_time} {writeup.id} " +
+                          f"{writeup.entityId}")
+    archive = syn.store(new_project)
+    permissions.set_entity_permissions(syn, archive, admin, "admin")
+    synapseutils.copy(syn, writeup.entityId, archive.id)
+
+
+def _validate_ent_type(ent):
+    """Helper function: check entity type."""
+
+    try:
+        if not isinstance(ent.entity, entity.Project):
+            ent_type = re.search(
+                r"entity\.(.*?)'", str(type(ent.entity))).group(1)
+            return f"Submission should be a Synapse project, not a {ent_type}."
+    except AttributeError:
+        return "Unknown entity type; please submit a Synapse project."
+    else:
+        return ""
+
+
+def _validate_project_contents(proj, challenge):
```

Comment:
I see.  This doesn't actually validate the content of the project, rather it just validates that the submission isn't the a particular project.